addon/nodes/paragraphWithConfig.ts

```diff
@@ -1,16 +1,102 @@
 import { NodeSpec } from 'prosemirror-model';
-import { paragraph } from './paragraph';
+import { getRdfaAttrs } from '@lblod/ember-rdfa-editor';
+import { NON_BLOCK_NODES } from '@lblod/ember-rdfa-editor/utils/_private/constants';
+import { optionMapOr } from '../utils/_private/option';
 
-interface paragraphConfig {
+export type ParagraphDataAttributes = {
+  'data-indentation-level': number;
+  'data-sub-type'?: string;
+};
+
+export type ParagraphNodeSpec = NodeSpec & { subType: string };
+
+let BLOCK_SELECTOR = '';
+NON_BLOCK_NODES.forEach(
+  (tag) => (BLOCK_SELECTOR = `${BLOCK_SELECTOR}${tag}, `),
+);
+BLOCK_SELECTOR = `:not(${BLOCK_SELECTOR.substring(
+  0,
+  BLOCK_SELECTOR.length - 2,
+)})`;
+
+const BASE_PARAGRAPH_TYPE = 'paragraph';
+const matchingSubType = (node: HTMLElement, subType: string) => {
+  // basic paragraph has no subtype in its dataset and an empty subType
+  const isBasicParagraph = node.dataset.subType === undefined && subType === '';
+  return isBasicParagraph || node.dataset.subType === subType;
+};
+
+export interface ParagraphConfig {
   content?: string;
   marks?: string;
   group?: string;
+  /* A mandatory subType name to separate different kinds of paragraphs in toDom functions.
+  For the basic paragraph this is the empty string and will not be added to dataset */
+  subType: string;
 }
 
-export const paragraphWithConfig: (config?: paragraphConfig) => NodeSpec = ({
-  content = paragraph.content,
-  marks = paragraph.marks,
-  group = paragraph.group,
-} = {}) => {
-  return { ...paragraph, content: content, marks: marks, group: group };
+export const paragraphWithConfig: (
+  config: ParagraphConfig,
+) => ParagraphNodeSpec = (config) => {
+  const name = config.subType !== '' ? config.subType : BASE_PARAGRAPH_TYPE;
+  return {
+    name: name,
```

Comment:
Is the name attribute just for debugging purposes?

Reply:
I think we don't use the name anywhere in code. I do often use it for debugging indeed.
So it is not required, but seemed less clean to keep it empty, because there is nothing bad about it being a sensible name I think? :)

Reply:
I agree :)